src/software/thunderscope/robot_diagnostics/handheld_device_manager.py

```diff
@@ -0,0 +1,341 @@
+        """Attempt to reinitialize a connection to a handheld device."""
+        self.__clear_handheld_device()
+        self.__initialize_handheld_device()
+
+    def __initialize_handheld_device(self) -> None:
+        """Attempt to initialize a connection to a handheld device.
+        The first handheld device that is recognized as valid will be used.
+        Valid handheld devices are any devices whose name has a matching DeviceConfig
+        in DiagnosticsConstants.HANDHELD_DEVICE_NAME_CONFIG_MAP.
+        """
+        device_config_map = DiagnosticsConstants.HANDHELD_DEVICE_NAME_CONFIG_MAP
+
+        for device in list_devices():
+            handheld_device = InputDevice(device)
+            if (
+                handheld_device is not None
+                and handheld_device.name in device_config_map
+            ):
+                with self.lock:
+                    self.handheld_device = handheld_device
+                    self.handheld_device_config = device_config_map[
+                        handheld_device.name
+                    ]
+
+                    self.handheld_device_connection_status_signal.emit(
+                        HandheldDeviceConnectionChangedEvent(
+                            HandheldDeviceConnectionStatus.CONNECTED,
+                            self.handheld_device.name,
+                        )
+                    )
+
+                    self.logger.debug("Successfully initialized handheld!")
+                    self.logger.debug(
+                        'Device name: "' + self.handheld_device.name + '"'
+                    )
+                    self.logger.debug("Device path: " + self.handheld_device.path)
+
+                return
+
+        self.__clear_handheld_device()
+        self.logger.debug("Failed to initialize handheld device, check USB connection")
+        self.logger.debug("Tried the following available devices:")
+        self.logger.debug(list(map(lambda d: InputDevice(d).name, list_devices())))
+
+    def __clear_handheld_device(self) -> None:
+        """Clear the handheld device and config fields by setting them to None,
+        and emit a disconnected notification signal.
+        """
+        with self.lock:
+            self.handheld_device = None
+            self.handheld_device_config = None
+
+        self.handheld_device_connection_status_signal.emit(
+            HandheldDeviceConnectionChangedEvent(
+                HandheldDeviceConnectionStatus.DISCONNECTED, None
+            )
+        )
+
+    def __event_loop(self) -> None:
+        """Loop that continuously reads and processes events from the connected handheld device."""
+        while True:
+            with self.lock:
+                self.__read_and_process_event()
+            time.sleep(DiagnosticsConstants.EVENT_LOOP_SLEEP_DURATION)
+
+    def __read_and_process_event(self) -> None:
+        """Try reading an event from the connected handheld device and process it."""
```

Comment:
this function should be documented as **not** being thread-safe (requires mutex locking)